pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -220,12 +264,153 @@ func (r *ReconcileNodeMaintenance) stopNodeMaintenance(nodeName string) error {
+		return err
+    }
+
 	return nil
 }
 
+func (r *ReconcileNodeMaintenance) invalidateLease(nodeName string) error {
+		if r.isLeaseSupported {
+				log.Info("Lease object supported,  invalidating lease")
+
+				nName := apitypes.NamespacedName{Namespace: LeaseNamespace, Name: nodeName}
+				lease := &coordv1beta1.Lease{}
+
+				if err := r.client.Get(context.TODO(), nName, lease); err != nil {
+					return err
+				}
+				lease.Spec.AcquireTime = nil
+				lease.Spec.LeaseDurationSeconds = nil;
+				lease.Spec.RenewTime = nil
+				lease.Spec.LeaseTransitions = nil
+
+				if err := r.client.Update(context.TODO(), lease); err != nil {
+					return err
+				}
+		}
+		return nil
+}
+
+func (r *ReconcileNodeMaintenance) obtainLease(node *corev1.Node, durationInSeconds int32) (reconcile.Result, error) {
+
+	lease, needUpdate, err := r.createOrGetExistingLease(node, durationInSeconds)
+
+	if err != nil {
+		log.Errorf("MaintOn: failed to create or get existing lease error=%v", err)
+		return reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}, err
+	}
+
+	if needUpdate {
+
+		log.Info("MaintOn: update lease - lease duration change requested")
+
+		if _, err = r.updateLease(node, lease, durationInSeconds); err != nil {
+				return reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+
+func makeInt32(val int32) (*int32) {
+	tmpVal := val
+	return &tmpVal
+}
+
+func makeTimeNow() (*metav1.MicroTime) {
+
+	timeNow := metav1.MicroTime{Time: time.Now()}
+	return &timeNow
+}
+
+
+func (r *ReconcileNodeMaintenance) createOrGetExistingLease(node *corev1.Node, durationInSeconds int32) (*coordv1beta1.Lease, bool, error) {
+	nodeName := node.ObjectMeta.Name
+
+	holderIdentity := LeaseHolderIdentity
+
+	owner := &metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       nodeName,
+		UID:        node.ObjectMeta.UID,
+	}
+
+	microTimeNow :=  makeTimeNow()
+
+	lease := &coordv1beta1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            nodeName,
+			Namespace:       LeaseNamespace,
+			OwnerReferences: []metav1.OwnerReference{*owner},
+		},
+		Spec: coordv1beta1.LeaseSpec{
+			HolderIdentity:       &holderIdentity,
+			LeaseDurationSeconds: makeInt32(durationInSeconds),
+			AcquireTime:          microTimeNow,
+			RenewTime:            microTimeNow,
+			LeaseTransitions:     makeInt32(1),
+		},
+	}
+
+	if err := r.client.Create(context.TODO(), lease); err != nil {
+		if errors.IsAlreadyExists(err) {
+
+				nName := apitypes.NamespacedName{Namespace: LeaseNamespace, Name: nodeName}
+
+				if err := r.client.Get(context.TODO(), nName, lease); err != nil {
+					return  nil, false, err
+				}
+				return lease, true, nil
+		}
+		return nil, false, err
+	}
+	return lease, false,  nil
+}
+
+func (r *ReconcileNodeMaintenance) updateLease(node *corev1.Node, lease *coordv1beta1.Lease, durationInSeconds int32) (*coordv1beta1.Lease, error) {
```

Comment:
We need a set of tests for `updateLease()` that are structured in a similar way to `checkLeaseSupported()`
Use `TableTest` if you wish, but it must be possible to add new test cases (input + expected output) without affecting existing ones, and the code executing the tests must not have conditional paths.
You may need to change the function to be able to accept a FakeClient.

Reply:
done